Alerting: Expose NoData and Error states for Alert Instances when remapped

joeblubaugh opened the pull request:
Alert Rules can be configured to use the "Alerting" or "OK" states when
rule evaluations result in Error or NoData conditions. This change
tracks the difference between, for example, Alerting, Alerting (NoData),
and Alerting (Error) by adding additional states to the model.

These additional states behave like their "base types" when deciding
whether to send alerts to the alertmanager, and how to display the
states in the UI, while adding additional context.

Fixes #42606

Signed-off-by: Joe Blubaugh <joe.blubaugh@grafana.com>

**Special notes for your reviewer**:

So far, the PR does not have tests, because I wanted to get reviewer opinions on a few things I've done here before updating tests to handle the new types. I've added comments in the attached review that raise the questions I'd like to answer and see if the approach needs to change.

joeblubaugh committed:
Alerting: Expose NoData and Error states for Alert Instances when remapped

Alert Rules can be configured to use the "Alerting" or "OK" states when
rule evaluations result in Error or NoData conditions. This change
tracks the difference between, for example, Alerting, Alerting (NoData),
and Alerting (Error) by adding additional states to the model.

These additional states behave like their "base types" when deciding
whether to send alerts to the alertmanager, and how to display the
states in the UI, while adding additional context.

Fixes #42606

Signed-off-by: Joe Blubaugh <joe.blubaugh@grafana.com>

joeblubaugh commented on public/app/features/alerting/unified/components/DynamicTable.tsx:
```diff
@@ -197,7 +197,6 @@ const getStyles = <T extends unknown>(
     `,
     bodyCell: css`
       overflow: hidden;
-      word-break: break-all;
```
I had to remove this becuase "break-all" would create: 

```
Alerting (Noda
ta)
```

whereas the default "normal" setting does

```
Alerting
(Nodata)
```

This line was added in a change specific to the alerting instances table, so I feel ok removing it for an alerting-instances-table improvement. I tested this on a mobile viewport and the table overflow was indistinguishable whatever the word-break setting was. https://github.com/grafana/grafana/commit/e5f6ab7fd2cb34fe21b88c6c53e86e1dd375024d

joeblubaugh reviewed (commented):
For later reviewers: there are two components in this change, and the connective tissue there is the text that is set in the annotations table, and in the instance state response for APIs. The valid values went from:

```
NoData
Error
Normal
Firing
Pending
```
to
```
NoData
Error
Normal
Normal (NoData)
Normal (Error)
Firing
Firing (NoData)
Firing (Error)
Pending
Pending (NoData)
Pending (Error)
```

There are conversions functions throughout that make it clear that, for example, all the Firing states are treated the same way that `eval.Alerting` was treated before. The main area where that feels rough and could use some love is in the frontend code that handles filtering the instance table by state. This used to be a simple 1:1 value mapping, but now it's more like 1:3 where filtering by "Normal" gives you Normal NormalNoData and NormalError. The code for that feels ugly right now.

joeblubaugh commented on pkg/services/ngalert/state/manager.go:
```diff
@@ -234,27 +236,39 @@ func (st *Manager) Put(states []*State) {
-	switch {
-	case state == ngModels.InstanceStateFiring:
+	switch state {
+	case ngModels.InstanceStateFiring,
```
replace with `isFiring`, etc

joeblubaugh committed:
Commit the Go type changes as suggested by Yuri

gillesdemey commented on public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:
```diff
@@ -79,7 +79,7 @@ function filterAlerts(
   if (alertInstanceState) {
     filteredAlerts = filteredAlerts.filter((alert) => {
-      return alert.state === alertInstanceState;
+      return alert.state.split(' ')[0] === alertInstanceState;
```
I think a function like `matchMappedAlertStateToInitialAlertState()` would make more sense instead of `split(' ')[0]`, would you mind adding that one?

joeblubaugh replied:
Sure, happy to encapsulate this. Are you fine with an in-line named function or do you think it should exist at the top-level of the component like `filterAlerts` does, or at the module level?

gillesdemey commented on public/app/features/alerting/unified/components/rules/AlertInstanceStateFilter.tsx:
```diff
@@ -10,10 +11,19 @@ interface Props {
-    label: value,
-    value,
-  }));
+  // Group our state options together by their "prefix", like Normal from
```
See my comment on `matchMappedAlertStateToInitialAlertState()` below, I think we can use that one to select only the base states.

joeblubaugh committed:
Implement Alert Instance State as a struct instead of a string

Using a struct makes it simpler to separate the "reason" for a change
from the derived state that generates alerts sent to the alertmanager.

Signed-off-by: Joe Blubaugh <joe.blubaugh@grafana.com>

joeblubaugh commented:
@yuri-tceretian I redid this as a struct implementation with xorm and json conversion code.  PTAL

yuri-tceretian commented on pkg/services/ngalert/models/instance.go:
```diff
@@ -11,15 +13,121 @@ type AlertInstance struct {
+// The state's Reason is why the state has the Type. A common case is State
+type InstanceState struct {
+	Type   InstanceStateType  `json:"type"`
+	Reason InstanceReasonType `json:"reason"`
```
I wonder if we can make reason optional (empty string or nil reference) and get rid of Normal.

yuri-tceretian commented on pkg/services/ngalert/models/instance.go:
```diff
@@ -43,12 +151,32 @@ func (i InstanceStateType) IsValid() bool {
+type InstanceReasonType string
+
+const (
+	InstanceReasonNormal = "Normal"
```
What do you think about renaming to to `None` or getting rid completely

yuri-tceretian commented on pkg/services/ngalert/schedule/compat.go:
```diff
@@ -50,11 +49,11 @@ func stateToPostableAlert(alertState *state.State, appURL *url.URL) *models.Post
-	if alertState.State == eval.Error {
+	if alertState.State.Type == ngModels.InstanceStateError {
 		return errorAlert(nL, nA, alertState, urlStr)
 	}
```
I wonder if we can add reason to alert annotations if it is not `Normal`

yuri-tceretian commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -70,8 +72,14 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
+				return errors.New("Bad argument")
+			}
+
+			addToQuery(` AND current_state = ?`, string(stateBytes))
```
what if I need all Alerting states with any reason?

yuri-tceretian reviewed (commented):
I think that it would be better to add a new column `reason`. 
I thought more about the flat structure, I.e. keep State and Reason as members of `State` struct. I am ok with this approach as well.

joeblubaugh commented:
> keep State and Reason as members of State struct. I am ok with this approach as well.
@yuri-tceretian when you write that, do you mean:
```go
type State {
  CurrentState InstanceStateType
  CurrentReason InstanceReasonType
}
```
or what I have in the PR with a struct that joins them together? If you mean making them both fields of the `State` struct, what do you think the benefit of that is?
